ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/DownloadHeadersStep.java

```diff
@@ -85,22 +92,57 @@ private CompletableFuture<List<BlockHeader>> downloadHeaders(final SyncTargetRan
               protocolSchedule,
               protocolContext,
               ethContext,
+              synchronizerConfiguration,
               range.getEnd(),
               range.getSegmentLengthExclusive(),
               validationPolicy,
               metricsSystem)
           .run();
     } else {
       LOG.debug("Downloading headers starting from {}", range.getStart().getNumber());
-      return GetHeadersFromPeerByHashTask.startingAtHash(
-              protocolSchedule,
-              ethContext,
-              range.getStart().getHash(),
-              range.getStart().getNumber(),
-              headerRequestSize,
-              metricsSystem)
-          .run()
-          .thenApply(PeerTaskResult::getResult);
+      if (synchronizerConfiguration.isPeerTaskSystemEnabled()) {
+        return ethContext
+            .getScheduler()
+            .scheduleServiceTask(
+                () -> {
+                  GetHeadersFromPeerTask task =
+                      new GetHeadersFromPeerTask(
+                          range.getStart().getHash(),
+                          range.getStart().getNumber(),
+                          headerRequestSize,
+                          0,
+                          GetHeadersFromPeerTask.Direction.FORWARD,
+                          protocolSchedule);
+                  PeerTaskExecutorResult<List<BlockHeader>> taskResult =
+                      ethContext.getPeerTaskExecutor().execute(task);
+                  if (taskResult.responseCode() != PeerTaskExecutorResponseCode.SUCCESS
+                      || taskResult.result().isEmpty()) {
+                    return CompletableFuture.failedFuture(
+                        new RuntimeException("Unable to download headers for range " + range));
+                  }
+                  final AtomicReference<BlockHeader> highestBlockHeader =
+                      new AtomicReference<>(taskResult.result().get().getFirst());
+                  for (BlockHeader blockHeader : taskResult.result().get()) {
+                    if (highestBlockHeader.get().getNumber() < blockHeader.getNumber()) {
+                      highestBlockHeader.set(blockHeader);
+                    }
+                  }
+                  taskResult
+                      .ethPeer()
+                      .ifPresent((peer) -> peer.chainState().update(highestBlockHeader.get()));
```

Comment:
Updating the chainstate should be in the peer task so it is updated when anything uses that like in the peer system.